src/Workspaces/Core/Portable/LanguageServices/SyntaxFactsService/AbstractSyntaxFactsService.cs

```diff
@@ -79,7 +83,26 @@ protected static void Intern(StringTable stringTable, ArrayBuilder<string> build
             }
         }
 
-        public abstract bool TryGetDeclaredSymbolInfo(StringTable stringTable, SyntaxNode node, out DeclaredSymbolInfo declaredSymbolInfo);
+        public static string GetMetadataAritySuffix(int arity)
+        {
+            Debug.Assert(arity > 0);
+            return (arity <= s_aritySuffixesOneToNine.Length)
+                ? s_aritySuffixesOneToNine[arity - 1]
+                : string.Concat(GenericTypeNameManglingString, arity.ToString(CultureInfo.InvariantCulture));
+        }
+
+        public abstract bool TryGetDeclaredSymbolInfo(StringTable stringTable, SyntaxNode node, string rootNamespace, out DeclaredSymbolInfo declaredSymbolInfo);
+
+        /// <summary>
+        /// If <paramref name="node"/> is an extension method declaration, this will return true, otherwise it returns false.
+        /// If the return value is true and <paramref name="targetTypeName"/> is null, then it means this is a "complex" method
+        /// (as described at <see cref="SyntaxTreeIndex.ExtensionMethodInfo"/>).
+        /// </summary>
+        public abstract bool TryGetTargetTypeName(SyntaxNode node, out string targetTypeName);
+
+        public abstract bool TryGetAliasesFromUsingDirective(SyntaxNode node, out ImmutableArray<(string aliasName, string name)> aliases);
+
+        public abstract string GetRootNamespace(CompilationOptions compilationOptions);
```

Comment:
i'm not really a fan of these jammed onto this existing type.  or... if they do ned to be here, do they need to be public?  can they not be internal  and only used by the subclass or by the TryGetDeclaredSymbolInfo implementation?

Reply:
hmm, I couldn't think of a straightforward way to do what you suggested (and not sure if it worth the effort to do so to be honest, since this service is only used when constructing SyntaxTreeIndex). Any suggestions on a good way to achieve this?

Reply:
not a problem then.  #resolved.